crates/containerd-shim-wasm/src/sandbox/sync.rs

```diff
@@ -0,0 +1,117 @@
+use std::cell::OnceCell;
+use std::sync::{Arc, Condvar, Mutex};
+use std::time::Duration;
+
+/// A cell where we can wait (with timeout) for
+/// a value to be set
+pub struct WaitableCell<T> {
+    inner: Arc<WaitableCellImpl<T>>,
+}
+
+struct WaitableCellImpl<T> {
+    // Ideally we would just use a OnceLock, but it doesn't
+    // have the `wait` and `wait_timeout` methods, so we use
+    // a Condvar + Mutex pair instead.
+    // We can't guard the OnceCell **inside** the Mutex as
+    // that would produce ownership problems with returning
+    // `&T`. This is because the mutex doesn't know that we
+    // won't mutate the OnceCell once it's set.
+    mutex: Mutex<()>,
```

Comment:
Initially, it seems strange to me that the `Mutex` type isn't really guarding anything. After I read the comment, I understand that we can't guard the `OnceCell` due to the use of `&T` within the OnceCell issue. 

I am wondering if we can change it to `cell: Mutex<Option<T>>` instead of `OnceCell<T>`?

Reply:
The problem with that is that we wouldn't be able to return a reference to it's content. The general use case for a mutex doesn't know that once set the content can't change, and thus it's safe to share a reference to it with the world.
We can work around this issue if we require `Clone` and we return a `clone` of the type. That would work for the `(u32, DateTime)` exit result. It would make the `WaitableCell` less generic, but that might be fine, it's not like we are using it for anything else.

I am also considering whether I should just remove the `WaitableCell` and replace it with a `sync::OnceCell` from the `once-cell` crate (which is the inspiration for the `OnceCell` and `OnceLock` of the rust std lib).
Unfortunately when the std lib adopted them, they didn't include the `wait` method.
The main problem with this is that while they have a `wait` method, they don't have a `wait_timeout` method. The only place where we use timeouts in the wait is on tests, and it might be fine to just spin a thread in those cases.

Reply:
> I am wondering if we can change it to cell: Mutex<Option<T>> instead of OnceCell<T>?

In short, IIUC, no, that wouldn't solve the issue unless we require `Clone`.
But I'm happy to discuss further as I might be missing something.